adding bokehjs known issues, v2.3 (#11014)
bokeh/bokeh: sphinx/source/docs/user_guide/bokehjs.rst

    plt.show(plt.gridplot([[p1, p2], [p3, p4]], {plot_width:350, plot_height:350}));

.. _userguide_bokehjs_differences:

> Usage : Comparison to Python

Reviewer: I believe there is an extra whitespace after "Usage". Also: I feel like "comparison" is not the optimal word here - I'd prefer something like differences. How about:

> Differences to Python library

----------------------------

While the BokehJS models generally match the Python equivalents, there are some cases

> where usage can differ significantly between BokehJS and the Python equivalent. Some

Reviewer: ```suggestion
where usage can differ significantly between BokehJS and the Python library. Some
```

instances we have identified are listed below.

.. _userguide_bokehjs_differences_titles:

Updating figure titles
~~~~~~~~~~~~~~~~~~~~~~

Calling ``var plot = new Bokeh.Plotting.figure({title:"some title"})`` replaces plot.title (a ``Bokeh.Title()``
object by default) with the provided string. This makes it so the title cannot be dynamically updated later.
In Python, the equivalent approach casts the string to the underlying Title() object.

To preserve the ability to update the title later, use the following sequence instead:

> ``var plot = new Bokeh.Plotting.figure()``

Reviewer: For code blocks, please use the `code-block` directive:

  .. code-block:: javascript
    var plot = new Bokeh.Plotting.figure()
    plot.title.text = "some title"

``plot.title.text = "some title"``

.. _userguide_bokehjs_differences_gridplots:

> Using gridplot layouts

Reviewer: Maybe we can be a little more specific with this headline:

```suggestion
Using ``sizing_mode`` in gridplot layouts
```

~~~~~~~~~~~~~~~~~~~~~~

``Bokeh.Plotting.gridplot()`` does not inherit ``sizing_mode`` from the underlying plots. To match the expected
behavior from the Python equivalent, for example when using ``sizing_mode: 'stretch_width'``, the ``sizing_mode``
must be explicitly stated for both the ``figure()`` and the ``gridplot()`` objects.

.. _userguide_bokehjs_differences_range_padding:

> Adjusting ``range_padding``

Reviewer: An idea to make the headlie more specific:
```suggestion
Making axis range properties persitent
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~

Setting ``plot.x_range.start`` and ``plot.x_range.end``, or ``plot.x_range.range_padding`` to adjust whitespace around

> glyphs on X axis is not persistent across resets. To match expected behavior from Python output, where the preferences

Reviewer: ```suggestion
glyphs on an axis is not persistent across resets. To match expected behavior from Python output, where the preferences
```

> will hold across figure resets, set both ``plot.x_range.range_padding`` and ``plot.x_range._initial_range_padding``.

Reviewer: This is quite a long complex sentence. How about splitting it up:

> In the Python library, the values for these properties are persistent when a figure is reset. To achieve similar behavior in BokehJS, set both ``plot.x_range.range_padding`` and ``plot.x_range._initial_range_padding``.

> (Same for Y axis)

Reviewer: Instead of adding the Y axis in parentheses, I'd suggest talking about "axes" from the beginning (meaning both X and Y axes).

.. _userguide_bokehjs_differences_palette:

> Palette for ``Bokeh.Plotting`` objects

Reviewer: ```suggestion
Palettes for ``Bokeh.Plotting`` objects
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The color palette is available under ``Bokeh.Charts`` but not under ``Bokeh.Plotting``; to use with ``Bokeh.Plotting`` objects such as ``Bokeh.Plotting.image()``, the palette argument must be provided as a color string array, rather than as a string.

Reviewer: The semicolon could be a period:
```suggestion
The color palette is available under ``Bokeh.Charts`` but not under ``Bokeh.Plotting``. To use it with ``Bokeh.Plotting`` objects such as ``Bokeh.Plotting.image()``, the palette argument must be provided as a color string array, rather than as a string.
```

.. _userguide_bokehjs_differences_name_field:

Setting glyph names for ``HoverTool``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Calling ``plot.line([0,1],[0,1],{name:“the_name”})`` sets the name under ``renderer.glyph.name``, rather than

> ``renderer.name``. The latter is used for HoverTool, under the special field ``$name``. To enable this special field in the HoverTool, consider passing the glyph name to the renderer:

Reviewer: We can be more direct here and just make a recommendation on what to do (instead of recommending to consider doing something): 
```suggestion
``renderer.name``. The latter is used for HoverTool, under the special field ``$name``. To enable this special field in the HoverTool, pass the glyph name to the renderer:
```

> ``plot.renderers[0].name = plot.renderers[0].glyph.name;``

Reviewer: ```suggestion
.. code-block:: javascript
    plot.renderers[0].name = plot.renderers[0].glyph.name;
```

.. _userguide_bokehjs_differences_div_widget:

Using ``Div()`` in a ``gridplot()``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

A ``Div()`` instance created by calling ``var mydiv = new Bokeh.Widgets.Div()`` does not include ``mydiv.toolbar.tools``
to facilitate using in a custom ``Bokeh.Plotting.gridplot()``; or ``gridplot()`` fails to check whether the property
exists before accessing. This throws ``Uncaught TypeError: Cannot read property 'tools' of undefined``

To use a ``Div()`` inside a ``gridplot()``, make sure to first set the div's toolbar property, as follows:

> ``mydiv.toolbar = {tools:[]}``

Reviewer: ```suggestion
.. code-block:: javascript
    mydiv.toolbar = {tools:[]}
```

to adjust whitespace around glyphs on an X axis (or the corresponding properties for a
Y axis) is not persistent across resets. In the Python library, the values for these
properties are persistent when a figure is reset. To achieve similar behavior in BokehJS,
set both ``plot.x_range.range_padding`` and ``plot.x_range._initial_range_padding``.

.. _userguide_bokehjs_differences_palette:

Palettes for ``Bokeh.Plotting`` objects
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The color palette is available under ``Bokeh.Charts`` but not under ``Bokeh.Plotting``.
To use it with ``Bokeh.Plotting`` objects such as ``Bokeh.Plotting.image()``, the
palette argument must be provided as a color string array, rather than as a string.

.. _userguide_bokehjs_differences_name_field:

Setting glyph names for ``HoverTool``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Calling ``plot.line([0,1],[0,1],{name:“the_name”})`` sets the name under ``renderer.glyph.name``,
rather than ``renderer.name``. The latter is used for HoverTool, under the special field
``$name``. To enable this special field in the HoverTool, pass the glyph name to the renderer:

.. code-block:: javascript

>     plot.renderers[0].name = plot.renderers[0].glyph.name;

Reviewer 2: This is fixed in PR #11018.

Author: @mattpap Awesome! That will be nice to have built in. For now tho, since #11018 is pointing to branch-2.4, I assume we'd still want an update to the live 2.3 docs. That is, assuming @bryevdv's position on redeploying docs when this PR is ready still stands?

Reviewer 3: To be honest @oviquezr I am having some second thoughts about this as it is currently formulated. What I would actually personally like to limit a "Differences between Python and JS" section to is *intentional* differences, not bugs. The best place to document bugs is in the issue tracker (not least because they will accurately represent status by being open or closed)

I think what I might prefer for this is to:

* Make individual issues on GitHub for all these bugs (including any workarounds)
* Link the *issues* in the tracker under a section called "Known issues"

That would:

* reduce the docs upkeep burden (the issues will always be up to date, necessarily)
* reduce duplication between the tracker and docs
* leave a future "differnces" section open to be clearly for describing intentional differences

I would be *much* more comfortable with that. 

cc @tcmetzger

Reviewer 3: FYI there is a Bokeh sphinx extension for linking issues:
```
:bokeh-issue:`1234`
```

.. _userguide_bokehjs_differences_div_widget:

Using ``Div()`` in a ``gridplot()``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

A ``Div()`` instance created by calling ``var mydiv = new Bokeh.Widgets.Div()``
does not include ``mydiv.toolbar.tools`` to facilitate using in a custom
``Bokeh.Plotting.gridplot()``; or ``gridplot()`` fails to check whether the
property exists before accessing. This throws ``Uncaught TypeError: Cannot read
property 'tools' of undefined``

To use a ``Div()`` inside a ``gridplot()``, make sure to first set the div's
toolbar property, as follows:

.. code-block:: javascript

>     mydiv.toolbar = {tools:[]}

Reviewer 2: bokeh's and bokehjs' match at API level, however bokeh's implementation is more permissive, breaking the API promise.

Reviewer 2: It's bad advice to tell users to do this kind of patching, because this taps into bokehjs' internal implementation details, and those can change at any time, breaking users' code.

Author: I agree the patch is suboptimal. If there is a better solution, I'd be happy to update these docs edits and use it in my own code! The proposed workaround is what did it for me, though. It being a single-line patch, it is also clean enough that I could tag it for review in my code, should the inner workings change in future Bokeh updates.
It sure would have been nice to find something like it in the docs as I was translating my original python plots to JS, which is why I'm recommending it here (took me some time of hunting to get the JS translation working). If there is no better working solution at the moment, perhaps we can use more direct language to make it clear that this is a temporary workaround?

Reviewer 3: @oviquezr I guess the point is that information lives forever on the internet, even when it goes stale and out of date, and for this reason we should simply refrain from documenting some workaround, if they are will cause more harm than good, in the future. Some things are simply bug to fix, when there are resources to address them, and be left at that.

Author: @bryevdv @mattpap  That I can most definitely live with! I'll move the `Div()` bit off this docs update and into its own issue in the tracker. I'm still, slowly, working my way through converting the original list from the discourse topic into hopefully helpful bug reports...
